src/proof_system/prover.rs

```diff
@@ -0,0 +1,322 @@
+        &mut self.cs
+    }
+    /// Preprocesses the underlying constraint system
+    pub fn preprocess(&mut self, commit_key: &ProverKey) -> PreProcessedCircuit {
+        self.cs
+            .preprocess(commit_key, &mut self.preprocessed_transcript)
+    }
+}
+
+// Draft 5: Lets just have the Prover wrap the composer functions
+// Cons of this approach is that we are duplicating. So might throw this away and go for something similar to Draft 4
+// In this idea, we have one composer per prover. So once we instantiate a Prover, you should not change the Composer/Preprocessed circuit
+
+impl Default for Prover {
+    fn default() -> Prover {
+        Prover::new()
+    }
+}
+
+impl Prover {
+    /// Creates a new prover object
+    pub fn new() -> Prover {
+        Prover {
+            preprocessed_circuit: None,
+            cs: StandardComposer::new(),
+            preprocessed_transcript: Transcript::new(b"plonk"),
+        }
+    }
+
+    /// Split `t(X)` poly into 3 degree `n` polynomials.
+    pub fn split_tx_poly(
+        &self,
+        n: usize,
+        t_x: &Polynomial,
+    ) -> (Polynomial, Polynomial, Polynomial, Polynomial) {
+        (
+            Polynomial::from_coefficients_vec(t_x[0..n].to_vec()),
+            Polynomial::from_coefficients_vec(t_x[n..2 * n].to_vec()),
+            Polynomial::from_coefficients_vec(t_x[2 * n..3 * n].to_vec()),
+            Polynomial::from_coefficients_vec(t_x[3 * n..].to_vec()),
+        )
+    }
+    /// Computes the quotient opening polynomial.
+    fn compute_quotient_opening_poly(
+        n: usize,
+        t_1_poly: &Polynomial,
+        t_2_poly: &Polynomial,
+        t_3_poly: &Polynomial,
+        t_4_poly: &Polynomial,
+        z_challenge: &Scalar,
+    ) -> Polynomial {
+        // Compute z^n , z^2n , z^3n
+        let z_n = z_challenge.pow(&[n as u64, 0, 0, 0]);
+        let z_two_n = z_challenge.pow(&[2 * n as u64, 0, 0, 0]);
+        let z_three_n = z_challenge.pow(&[3 * n as u64, 0, 0, 0]);
+
+        let a = t_1_poly;
+        let b = t_2_poly * &z_n;
+        let c = t_3_poly * &z_two_n;
+        let d = t_4_poly * &z_three_n;
+        let abc = &(a + &b) + &c;
+        &abc + &d
+    }
+    /// Convert variables to their actual witness values.
+    pub(crate) fn to_scalars(&self, vars: &[Variable]) -> Vec<Scalar> {
+        vars.par_iter().map(|var| self.cs.variables[var]).collect()
+    }
+
+    pub(crate) fn reset(&mut self) {
+        self.cs.w_l.clear();
+        self.cs.w_r.clear();
+        self.cs.w_o.clear();
+        self.cs.w_4.clear();
+
+        // Do not reset the Transcript as we need the previous messages exchanged during the preprocessing stage
+    }
+
+    /// Prove will compute the pre-processed polynomials and
+    /// produce a proof
+    /// We assume that the Prover struct has a composer
+    /// XXX: Accept a string, that we can key the transcript with?
+    pub fn prove(&mut self, commit_key: &ProverKey) -> Proof {
```

Comment:
Ye, it's not the best approach but I can't think on something better atm.

Accept a field on the function like: `transcript_seed: Option<&str>`

Reply:
Oh, I meant the whole approach is questionable to me.

This comment was left from a previous idea I had. In this idea, we can initialise the Prover and Verifier with  label and additionally seed the transcript if needed. Have pushed this a few minutes ago